File: test/CryptoTests.c
//
//  CryptoTests.c
//  yammer
//
//  Created by david on 12/2/15.
//  Copyright © 2015 combobulated. All rights reserved.
//

#include "CryptoTests.h"

#include "YMRSAKeyPair.h"
#include "YMX509Certificate.h"

YM_EXTERN_C_PUSH

typedef struct CryptoTest
{
    ym_test_assert_func assert;
    const void *context;
} CryptoTest;

void _X509GenerationTestRun(struct CryptoTest *test);

void CryptoTestsRun(ym_test_assert_func assert, const void *context)
{
    struct CryptoTest theTest = { assert, context };
    
    _X509GenerationTestRun(&theTest);
    ymerr(" _X509GenerationTestRun completed");
}

void _X509GenerationTestRun(struct CryptoTest *theTest)
{    
    YMRSAKeyPairRef keyPair = YMRSAKeyPairCreate();
    testassert(YMRSAKeyPairGenerate(keyPair), "generate failed");
    YMX509CertificateRef cert = YMX509CertificateCreate(keyPair);
    testassert(cert, "cert creation failed");
    YMRelease(cert);
    YMRelease(keyPair);
    
    int bits = 4096;
    keyPair = YMRSAKeyPairCreateWithModuloSize(bits, 65537);
    testassert(YMRSAKeyPairGenerate(keyPair), "generate(%d) failed",bits);
    cert = YMX509CertificateCreate(keyPair);
    testassert(cert, "cert creation failed");
    YMRelease(cert);
    YMRelease(keyPair);
}

YM_EXTERN_C_POP

File: osx/YammerTests/CryptoTests.m
//
//  CryptoTests.m
//  yammer
//
//  Created by david on 11/9/15.
//  Copyright © 2015 combobulated. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "YammerTests.h"

#include "YMBase.h"

#include "YMRSAKeyPair.h"
#include "YMX509Certificate.h"

@interface CryptoTests : XCTestCase

@end

@implementation CryptoTests

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testX509Generation {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
    
    YMRSAKeyPairRef keyPair = YMRSAKeyPairCreate();
    XCTAssert(YMRSAKeyPairGenerate(keyPair), @"generate failed");
    YMX509CertificateRef cert = YMX509CertificateCreate(keyPair);
    XCTAssert(cert, @"cert creation failed");
    YMRelease(cert);
    YMRelease(keyPair);
    
    keyPair = YMRSAKeyPairCreateWithModuloSize(4096, 65537);
    XCTAssert(YMRSAKeyPairGenerate(keyPair), @"generate(4096) failed");
    cert = YMX509CertificateCreate(keyPair);
    XCTAssert(cert, @"cert creation failed");
    YMRelease(cert);
    YMRelease(keyPair);
}

@end
